File: apps/ios/PhrenKit/Sources/PhrenKit/Markdown/ReleaseNotes.swift
import Foundation

/// The app's CHANGELOG.md, read for "What's new": one `## <version>` section
/// per release, `### New` / `### Improved` / `### Fixed` groups inside it,
/// one `- ` bullet per change (continuation lines indented). Anything that is
/// not a heading or a bullet — the preamble, "Earlier" prose — is kept as a
/// note on the section.
public struct ReleaseNotes: Equatable, Sendable {
    public struct Group: Equatable, Sendable, Identifiable {
        public let title: String
        public let items: [String]
        public var id: String { title }
    }
    public struct Release: Equatable, Sendable, Identifiable {
        public let version: String
        public let groups: [Group]
        public let notes: [String]
        public var id: String { version }
        public var isEmpty: Bool { groups.allSatisfy(\.items.isEmpty) && notes.isEmpty }
    }
    public let releases: [Release]

    public init(markdown: String) {
        var releases: [Release] = []
        var version: String?, groups: [Group] = [], title = "", items: [String] = [], notes: [String] = []
        func closeGroup() { if !items.isEmpty || !title.isEmpty { groups.append(Group(title: title, items: items)) }; items = []; title = "" }
        func closeRelease() { closeGroup(); if let version { releases.append(Release(version: version, groups: groups, notes: notes)) }; groups = []; notes = [] }
        for raw in markdown.prefix(262_144).components(separatedBy: "\n") {
            let line = raw.trimmingCharacters(in: .whitespaces)
            if raw.hasPrefix("## ") {
                closeRelease(); version = String(raw.dropFirst(3)).trimmingCharacters(in: .whitespaces)
            } else if raw.hasPrefix("### ") {
                closeGroup(); title = String(raw.dropFirst(4)).trimmingCharacters(in: .whitespaces)
            } else if line.hasPrefix("- ") {
                items.append(String(line.dropFirst(2)))
            } else if !line.isEmpty, version != nil, !raw.hasPrefix("#") {
                // A wrapped bullet continues the last item; anything else is a note.
                if raw.hasPrefix("  "), !items.isEmpty { items[items.count - 1] += " " + line }
                else if title.isEmpty { notes.append(line) }
            }
        }
        closeRelease()
        self.releases = releases
    }

    /// The section for one version — "0.0.6" matches "## 0.0.6" and
    /// "## 0.0.6 (build 61)" alike.
    public func release(for version: String) -> Release? {
        releases.first { $0.version == version || $0.version.hasPrefix(version + " ") }
    }
}

File: apps/android/phrenkit/src/main/kotlin/com/phren/kit/ReleaseNotes.kt
package com.phren.kit

/**
 * The app's CHANGELOG.md for "What's new" (ReleaseNotes.swift): one
 * `## <version>` section per release, `### ` groups inside it, one `- `
 * bullet per change with indented continuation lines. Other text in a
 * section is kept as a note.
 */
class ReleaseNotes(markdown: String) {
    data class Group(val title: String, val items: List<String>)
    data class Release(val version: String, val groups: List<Group>, val notes: List<String>) {
        val isEmpty: Boolean get() = groups.all { it.items.isEmpty() } && notes.isEmpty()
    }

    val releases: List<Release>

    init {
        val out = mutableListOf<Release>()
        var version: String? = null
        val groups = mutableListOf<Group>()
        var title = ""
        val items = mutableListOf<String>()
        val notes = mutableListOf<String>()
        fun closeGroup() { if (items.isNotEmpty() || title.isNotEmpty()) groups += Group(title, items.toList()); items.clear(); title = "" }
        fun closeRelease() { closeGroup(); version?.let { out += Release(it, groups.toList(), notes.toList()) }; groups.clear(); notes.clear() }
        for (raw in markdown.take(262_144).split("\n")) {
            val line = raw.trim(' ', '\t')
            when {
                raw.startsWith("## ") -> { closeRelease(); version = raw.drop(3).trim(' ', '\t') }
                raw.startsWith("### ") -> { closeGroup(); title = raw.drop(4).trim(' ', '\t') }
                line.startsWith("- ") -> items += line.drop(2)
                line.isNotEmpty() && version != null && !raw.startsWith("#") -> {
                    if (raw.startsWith("  ") && items.isNotEmpty()) items[items.lastIndex] = items.last() + " " + line
                    else if (title.isEmpty()) notes += line
                }
            }
        }
        closeRelease()
        releases = out
    }

    /** "0.0.6" matches "## 0.0.6" and "## 0.0.6 (build 61)" alike. */
    fun release(version: String): Release? = releases.firstOrNull { it.version == version || it.version.startsWith("$version ") }
}
